src/flat_capi/engine_pretrain.py

```diff
@@ -0,0 +1,250 @@
+from __future__ import annotations
+
+from typing import Iterable, Callable
+
+import torch
+import wandb
+from torch import Tensor, nn
+
+from flat_capi.utils import misc
+
+
+def _get_do_teacher(model) -> Callable:
+    """Get the teacher forward function, preferring compiled version if present."""
+    return getattr(model, "do_teacher_compiled", do_teacher)
+
+
+def _get_do_student(model) -> Callable:
+    """Get the student loss function, preferring compiled version if present."""
+    return getattr(model, "do_student_compiled", do_student)
+
+
+def do_student(
+    model: nn.Module,
+    images: Tensor,
+    predict_indices: Tensor,
+    visible_indices: Tensor,
+    target: Tensor,
+    temp: float,
+) -> Tensor:
+    """Compute the CAPI student loss against teacher targets.
+
+    Args:
+        model: Student-teacher wrapper.
+        images: Input batch [B, C, H, W] (or channels-as-time).
+        predict_indices: Absolute indices of tokens to predict.
+        visible_indices: Absolute indices of visible tokens.
+        target: Teacher assignments used as soft labels.
+        temp: Student softmax temperature.
+
+    Returns:
+        Mean loss tensor (double).
+    """
+    core = model.module if hasattr(model, "module") else model
+    amp_dtype = images.dtype if images.dtype in (torch.float16, torch.bfloat16) else torch.float32
```

Comment:
I think inputs are not supposed to be cast (see [here](https://docs.pytorch.org/docs/stable/amp.html#torch.autocast)). So you will be disabling amp by default.